packages/workspace/src/browser/workspace-service.ts

```diff
@@ -16,48 +16,108 @@ import { WorkspaceServer } from "../common";
 @injectable()
 export class WorkspaceService {
 
-    /**
-     * The current workspace root.
-     */
-    readonly root: Promise<FileStat>;
+    private rootWasResolvedAtStartup: Promise<boolean>;
 
     constructor(
         @inject(FileSystem) protected readonly fileSystem: FileSystem,
         @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
         @inject(WorkspaceServer) protected readonly server: WorkspaceServer
     ) {
-        this.root = this.server.getRoot().then(uri =>
-            this.validateRoot(uri)
-        );
-        this.root.then(root =>
-            watcher.watchFileChanges(new URI(root.uri))
-        );
+        this.rootWasResolvedAtStartup = this.rootResolved;
+    }
+
+    /**
+     * The promise which will resolve to the currently selected workspace root.
+     * This promise resolves to the workspace root file stat, if [rootResolved](WorkspaceService.rootResolved) is `true`.
+     */
+    get root(): Promise<FileStat> {
+        return new Promise(async (resolve, reject) => {
+            const root = await this.server.getRoot();
+            const validRoot = await this.isValidRoot(root);
+            if (validRoot) {
+                resolve(this.toFileStat(root!));
+            } else {
+                this.rootWasResolvedAtStartup = Promise.resolve(false);
```

Comment:
reassigning this field will not notify any listeners on the previously assigned promise. I think a Deferred Promise would simplify this code a bit. See `promise-util.ts` in core.